src/Commands/ToggleDone.ts

```diff
@@ -78,61 +89,34 @@ export const toggleLine = (line: string, path: string) => {
-1) Line stayed the same length: Checking & unchecking textbox that is not a task - cursor should reset.
-2) Line got longer:
-    a) List marker could have been added. Find it in new text: if cursor was at or right of where it was added, move the cursor right.
-    b) Empty checkbox could have been added. If cursor was after the list marker (in old or new), it should move right.
-    c) Done emoji and date could have been added to the end. Cursor should reset if 0, and stay end of line otherwise.
-    d) Recurring task could have been added to the beginning and done emoji and date added to the end. Current behavior adds so much to the offset to make this right.
-
-So cursor should be reset if 0, which includes being moved to new end if got shorter. Then might need to move right 2 or 3.
-*/
-export const calculateCursorOffset = (origCursorCh: number, line: string, toggledLine: string) => {
-    let newLineLen = toggledLine.length;
-    if (newLineLen <= line.length) {
-        // Line got shorter or stayed same length. Reset cursor to original position, capped at end of line.
-        return origCursorCh >= toggledLine.length ? newLineLen : origCursorCh;
-    }
-
-    // Special-case for done-date append, fixes #449
-    const doneDateLength = ' ✅ YYYY-MM-DD'.length;
-    if (toggledLine.match(TaskRegularExpressions.doneDateRegex) && newLineLen - line.length >= doneDateLength) {
-        newLineLen -= doneDateLength;
-    }
-
-    // Handle recurring tasks: entire line plus newline prepended. Fix for #449 above means appended done date treated correctly.
-    if (newLineLen >= 2 * line.length && toggledLine.search('.+\n.+') !== -1) {
-        return origCursorCh + newLineLen - line.length;
-    }
-
-    /* Line got longer, not a recurring task. Were the added characters before or after the cursor?
-     * At this point the line is at least a list item. Find the first list marker. */
-    const firstListItemChar = toggledLine.search(/[-*]/);
-    if (origCursorCh < firstListItemChar) {
-        // Cursor was in indentation. Reset to where it was.
-        return origCursorCh;
-    }
-
-    return origCursorCh + newLineLen - line.length;
+/**
+ * Computes the new position of the cursor, given its current position and an
+ * the suggested position within the inserted text.
+ *
+ * @note Assumes that the insertion occurs at column 0
+ *
+ * @param startPos The starting cursor position
+ * @param insertion The inserted text and suggested cursor position within that text
+ */
+export const getNewCursorPosition = (startPos: EditorPosition, insertion: EditorInsertion): EditorPosition => {
+    const line = insertion.moveTo?.line ?? 0;
+    const newCh = insertion.moveTo?.ch ?? startPos.ch;
+    const _min = (a: number, b: number) => (a < b ? a : b);
+    return {
+        line: startPos.line + line,
+        ch: _min(newCh, insertion.text[line].length), // This assumes that the inserted text is inserted at column 0
```

Comment:
What happens if that assumption is untrue?
Should there be any code to test that assumption?

Reply:
> I documented what could go wrong in [c177a16](https://github.com/obsidian-tasks-group/obsidian-tasks/commit/c177a16e6e7058e8a10abaff90713e13465aad91) , but I think going for correctness in all cases would add unnecessary complexity.

Brilliant. Thank you. Resolving.